File: truss-compiler-maven-plugin/src/main/java/truss/compiler/generator/CodeWriter.java
package truss.compiler.generator;

import org.apache.commons.lang.StringUtils;

public class CodeWriter {
    private boolean hadIndent = false;
    private int indent;
    private final StringBuilder sb = new StringBuilder();

    public void indent() {
        indent++;
    }

    public void unIndent() {
        indent--;
    }

    public void write(String format, Object... args) {
        if (!hadIndent && indent > 0) {
            sb.append(StringUtils.repeat(" ", indent * 4));
            hadIndent = true;
        }

        if (args != null && args.length > 0) {
            format = String.format(format, args);
        }

        sb.append(format);
    }

    public void writeln() {
        hadIndent = false;
        sb.append("\n");
    }

    public void writeln(String format, Object... args) {
        write(format, args);
        writeln();
    }

    @Override
    public String toString() {
        return sb.toString();
    }

    public void writeRaw(String validation) {
        sb.append(validation);
    }
}

File: Truss.Compiler.Generate/CodeWriter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using JetBrains.Annotations;

namespace Truss.Compiler.Generate {
    public class CodeWriter {
        private bool _hadIndent;
        private int _indent;
        private readonly StringBuilder _sb = new StringBuilder();

        public void Indent() {
            _indent++;
        }

        public void UnIndent() {
            _indent--;
        }

        public void Write(string format) {
            Write(format, null);
        }

        [StringFormatMethod("format")]
        public void Write(string format, params Object[] args) {
            if (!_hadIndent && _indent > 0) {
                _sb.Append(new string(' ', _indent * 4));
                _hadIndent = true;
            }

            if (args != null && args.Length > 0) {
                format = string.Format(format, args);
            }

            _sb.Append(format);
        }

        public void WriteLine() {
            _hadIndent = false;
            _sb.Append(Environment.NewLine);
        }

        public void WriteLine(string format) {
            WriteLine(format, null);
        }

        [StringFormatMethod("format")]
        public void WriteLine(string format, params Object[] args) {
            Write(format, args);
            WriteLine();
        }

        public override string ToString() {
            return _sb.ToString();
        }

        public void WriteRaw(string validation) {
            _sb.Append(validation);
        }
    }
}
